test/cpp/api/inference_mode.cpp

```diff
@@ -0,0 +1,276 @@
+  // If we have new intermediate tensor created in backend kernels, ops on it will dispatch to VariableType kernel
+  // unexpectedly.
+  // Erroring out in InplaceOrView kernel is much easier and less error prone.
+  ASSERT_THROWS_WITH(inplace_op(inference_tensor), // go through kernels: InplaceOrView(ERROR!), CPU
+    "inplace/view ops on inference tensor outside InferenceMode")
+  ASSERT_THROWS_WITH(view_op(inference_tensor), // go through kernels: InplaceOrView(ERROR!), CPU
+    "inplace/view ops on inference tensor outside InferenceMode")
+}
+
+// TODO: Currently requires_grad=false tensor behave exactly the same as requires_grad=true tensor
+//       inside dispatcher.
+//       In the ideal future requires_grad=false tensor don't go through VariableType kernel
+//       and we should add separate tests for require_grad=false tensor.
+TEST(InferenceModeTest, TestRequiresGradTrueTensorFunctionalOp) {
+  torch::Tensor a = torch::ones({1, 2, 3}).set_requires_grad(true);
+  torch::Tensor func_out;
+
+  {
+    c10::InferenceMode guard;
+    // Functional op on normal tensor in InferenceMode produces
+    // inference tensor as output.
+    func_out = functional_op(a);  // go through kernels: InplaceOrView, CPU
+    assert_tensor_dispatch_keys(func_out, /*is_inference_tensor=*/true);
+    ASSERT_FALSE(func_out.requires_grad());
+  }
+
+  // func_out should behave exactly the same as inference tensor.
+  torch::Tensor tmp = functional_op(func_out); // go through kernels: InplaceOrView(fallthrough), CPU
+  assert_tensor_dispatch_keys(tmp, /*is_inference_tensor=*/false);
+  ASSERT_FALSE(tmp.requires_grad());
+
+  ASSERT_THROWS_WITH(inplace_op(func_out), // go through kernels: InplaceOrView(ERROR!), CPU
+    "inplace/view ops on inference tensor outside InferenceMode")
+  ASSERT_THROWS_WITH(view_op(func_out), // go through kernels: InplaceOrView(ERROR!), CPU
+    "inplace/view ops on inference tensor outside InferenceMode")
+}
+
+TEST(InferenceModeTest, TestRequiresGradTrueTensorInplaceOp) {
+  torch::Tensor s = torch::ones({1, 2, 3}).set_requires_grad(true);
+  torch::Tensor a = s + 2;
+
+  {
+    c10::InferenceMode guard;
+
+    // Inplace ops on normal tensor produce normal tensors as output.
+    // In other words they have Autograd=true, InplaceOrView=true.
+    // It's straightforward since output is the same as input.
+    inplace_op(a);  // go through kernels: InplaceOrView, CPU
+    assert_tensor_dispatch_keys(a, /*is_inference_tensor=*/false);
+    ASSERT_TRUE(a.requires_grad());
+
+    // inplace -> inplace
+    inplace_op(a);  // go through kernels: InplaceOrView, CPU
+    assert_tensor_dispatch_keys(a, /*is_inference_tensor=*/false);
+    ASSERT_TRUE(a.requires_grad());
+
+    // inplace -> inplace -> view
+    torch::Tensor view_out = view_op(a);
+    assert_tensor_dispatch_keys(view_out, /*is_inference_tensor=*/false);
+    ASSERT_TRUE(view_out.requires_grad());
```

Comment:
Not only is the behavior equivalent, but if you were to run autograd on this result, it would work, right?

Reply:
No it wouldn't, because the InplaceOrView kernel only would version bump, it wouldn't also update the autograd metadata. So I wonder what actually would happen if you tried to call backward on the mutated tensor later.

Reply:
I'm not sure why you wouldn't mark the resulting view in this situation as `requires_grad=False`. The behavior here is consistent with `no_grad`, but this behavior is controversial https://github.com/pytorch/pytorch/issues/11390 also see #12502. Because inference mode is all about letting implications of things that happen in the inference mode region escape (in the name of performance), it doesn't make sense to me to preserve the old view behavior.

Actually, I might even go a step further, and say that inplacing a `requires_grad=True` tensor inside an inference mode block should set `requires_grad=False`. But I suppose we can cross that bridge later (certainly seems more annoying to implement).

Reply:
I suppose an argument for the current behavior is that InferenceMode is intended to be exactly semantically equivalent to `no_grad`, except that there are more error cases in some situations. However, I checked https://fb.quip.com/JFgeAL0arWCA and didn't see this explicated as an explicit design goal (though I could have missed it!)